### PR TITLE
Reset all results on epoch end

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/13554

### Does your PR introduce any breaking changes? If yes, please list them.

Epoch-end results will not be logged again in subsequent epochs.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @akihironitta

### CODE REVIEW

**Comment by awaelchli** [comment]
@carmocca I'm quite confused as to how the `on_step=False`, `on_epoch=True|False` affects whether the special `on_train_start` value is being logged or not. Is this expected?

```py
class BugModel(BoringModel):
    def on_train_start(self):
        print("ON_TRAIN_START")
        self.log("train_loss", 5000, on_step=False, on_epoch=True)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss, on_step=False, on_epoch=True)
        return loss

```

**Comment by carmocca** [comment]
> @carmocca I'm quite confused as to how the `on_step=False`, `on_epoch=True|False` affects whether the special `on_train_start` value is being logged or not. Is this expected?

This question is totally unrelated to the PR, but to answer it. It's because we differentiate results by the `{fx_name}.{logged_key}`, meaning that the values `self.log`ged in each hook don't get reduced together.

However, since they just get added to a simple dictionary (`logged_metrics`), and we only use `logged_key` as the key, the latter overwrites the former. It's a legacy design limitation

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]
